File: api/app/services/delivery.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import InventoryBalance, Order, OrderLine, StopDelivery, Trip, TripStop
from app.services.events import append_event


def _warehouse_for_trip(trip: Trip) -> uuid.UUID:
    return trip.warehouse_location_id


def apply_stop_delivery(
    db: Session,
    trip: Trip,
    stop: TripStop,
    item: StopDelivery,
    qty_delivered: float,
    status: str,
    reason_code: str | None,
    signature_url: str | None,
    photo_url: str | None,
) -> None:
    warehouse_id = _warehouse_for_trip(trip)
    line = db.get(OrderLine, item.order_line_id)
    if not line:
        raise ValueError("Order line missing")

    inv = db.execute(
        select(InventoryBalance).where(
            InventoryBalance.location_id == warehouse_id,
            InventoryBalance.sku_id == line.sku_id,
        ).with_for_update()
    ).scalar_one_or_none()
    if not inv:
        raise ValueError("Inventory row missing")

    qty = min(qty_delivered, item.qty_planned)
    reduce_reserved = min(qty, inv.reserved)
    inv.reserved -= reduce_reserved
    inv.on_hand -= qty

    item.qty_delivered = qty
    item.status = status
    item.reason_code = reason_code
    item.signature_url = signature_url
    item.photo_url = photo_url
    item.pod_captured_at = datetime.now(timezone.utc)

    if status == "delivered" and qty >= item.qty_planned:
        line.status = "delivered"
    elif status == "partial":
        line.status = "shipped"
    elif status == "failed":
        pass


def refresh_order_status(db: Session, order_id: uuid.UUID) -> None:
    lines = db.execute(select(OrderLine).where(OrderLine.order_id == order_id)).scalars().all()
    if not lines:
        return
    order = db.get(Order, order_id)
    if not order:
        return
    if all(ln.status == "delivered" for ln in lines):
        order.status = "delivered"
    elif any(ln.status == "delivered" for ln in lines):
        order.status = "shipped"


def mark_trip_completed_if_all_stops_done(db: Session, trip: Trip) -> None:
    stops = db.execute(select(TripStop).where(TripStop.trip_id == trip.id)).scalars().all()
    for stop in stops:
        deliveries = db.execute(select(StopDelivery).where(StopDelivery.trip_stop_id == stop.id)).scalars().all()
        for d in deliveries:
            if d.status == "pending":
                return
    trip.status = "completed"
    trip.completed_at = datetime.now(timezone.utc)

File: web/lib/server/delivery.ts
import type { Prisma, Trip, TripStop, StopDelivery } from "@prisma/client";
import { prisma } from "@/lib/prisma";

export async function applyStopDelivery(
  tx: Prisma.TransactionClient,
  trip: Trip,
  _stop: TripStop,
  item: StopDelivery,
  qtyDelivered: number,
  status: string,
  reasonCode: string | null,
  signatureUrl: string | null,
  photoUrl: string | null,
) {
  const warehouseId = trip.warehouseLocationId;
  const line = await tx.orderLine.findUnique({ where: { id: item.orderLineId } });
  if (!line) throw new Error("Order line missing");

  const inv = await tx.inventoryBalance.findFirst({
    where: { locationId: warehouseId, skuId: line.skuId },
  });
  if (!inv) throw new Error("Inventory row missing");

  const qty = Math.min(qtyDelivered, item.qtyPlanned);
  const reduceReserved = Math.min(qty, inv.reserved);

  await tx.inventoryBalance.update({
    where: { id: inv.id },
    data: {
      reserved: inv.reserved - reduceReserved,
      onHand: inv.onHand - qty,
    },
  });

  await tx.stopDelivery.update({
    where: { id: item.id },
    data: {
      qtyDelivered: qty,
      status,
      reasonCode: reasonCode ?? undefined,
      signatureUrl: signatureUrl ?? undefined,
      photoUrl: photoUrl ?? undefined,
      podCapturedAt: new Date(),
    },
  });

  if (status === "delivered" && qty >= item.qtyPlanned) {
    await tx.orderLine.update({ where: { id: line.id }, data: { status: "delivered" } });
  } else if (status === "partial") {
    await tx.orderLine.update({ where: { id: line.id }, data: { status: "shipped" } });
  }
}

export async function refreshOrderStatus(tx: Prisma.TransactionClient, orderId: string) {
  const lines = await tx.orderLine.findMany({ where: { orderId } });
  if (lines.length === 0) return;
  const order = await tx.order.findUnique({ where: { id: orderId } });
  if (!order) return;
  if (lines.every((ln) => ln.status === "delivered")) {
    await tx.order.update({ where: { id: orderId }, data: { status: "delivered" } });
  } else if (lines.some((ln) => ln.status === "delivered")) {
    await tx.order.update({ where: { id: orderId }, data: { status: "shipped" } });
  }
}

export async function markTripCompletedIfAllStopsDone(
  tx: Prisma.TransactionClient,
  tripId: string,
) {
  const stops = await tx.tripStop.findMany({ where: { tripId } });
  for (const stop of stops) {
    const deliveries = await tx.stopDelivery.findMany({ where: { tripStopId: stop.id } });
    for (const d of deliveries) {
      if (d.status === "pending") return;
    }
  }
  await tx.trip.update({
    where: { id: tripId },
    data: { status: "completed", completedAt: new Date() },
  });
}

/** Load trip with stops and deliveries (sorted stops) */
export async function loadTripFull(tripId: string) {
  const trip = await prisma.trip.findFirst({
    where: { id: tripId },
    include: {
      stops: {
        include: { deliveries: true },
        orderBy: { sequence: "asc" },
      },
    },
  });
  return trip;
}
